File: legacy/src/main/java/preponderous/viron/services/GridService.java
package preponderous.viron.services;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.web.client.RestTemplateBuilder;
import org.springframework.http.HttpMethod;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.HttpClientErrorException;
import org.springframework.web.client.RestTemplate;
import preponderous.viron.config.ServiceConfig;
import preponderous.viron.exceptions.ServiceException;
import preponderous.viron.models.Grid;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

@Service
@Slf4j
public class GridService {
    private final RestTemplateBuilder restTemplateBuilder;
    private final String baseUrl;

    @Autowired
    public GridService(RestTemplateBuilder restTemplateBuilder, ServiceConfig serviceConfig) {
        this.restTemplateBuilder = restTemplateBuilder;
        this.baseUrl = serviceConfig.getVironHost() + ":" + serviceConfig.getVironPort() + "/api/v1/grids";
    }

    public List<Grid> getAllGrids() {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Grid[]> response = restTemplate.getForEntity(baseUrl, Grid[].class);
            return Arrays.asList(response.getBody());
        } catch (Exception e) {
            log.error("Error getting all grids: {}", e.getMessage());
            throw new ServiceException("Failed to fetch all grids", e);
        }
    }

    public Optional<Grid> getGridById(int id) {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Grid> response = restTemplate.getForEntity(baseUrl + "/{id}", Grid.class, id);
            return Optional.ofNullable(response.getBody());
        } catch (HttpClientErrorException.NotFound e) {
            return Optional.empty();
        } catch (Exception e) {
            log.error("Error getting grid by id {}: {}", id, e.getMessage());
            throw new ServiceException("Failed to fetch grid by id: " + id, e);
        }
    }

    public List<Grid> getGridsInEnvironment(int environmentId) {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Grid[]> response = restTemplate.getForEntity(
                    baseUrl + "/environment/{environmentId}",
                    Grid[].class,
                    environmentId
            );
            return Arrays.asList(response.getBody());
        } catch (Exception e) {
            log.error("Error getting grids in environment {}: {}", environmentId, e.getMessage());
            throw new ServiceException("Failed to fetch grids in environment: " + environmentId, e);
        }
    }

    public Optional<Grid> getGridOfEntity(int entityId) {
        try {
            RestTemplate restTemplate = restTemplateBuilder.build();
            ResponseEntity<Grid> response = restTemplate.getForEntity(
                    baseUrl + "/entity/{entityId}",
                    Grid.class,
                    entityId
            );
            return Optional.ofNullable(response.getBody());
        } catch (HttpClientErrorException.NotFound e) {
            return Optional.empty();
        } catch (Exception e) {
            log.error("Error getting grid for entity {}: {}", entityId, e.getMessage());
            throw new ServiceException("Failed to fetch grid for entity: " + entityId, e);
        }
    }
}

File: legacy/src/main/python/preponderous/viron/services/gridService.py
# Copyright (c) 2024 Preponderous Software
# MIT License

import logging
from typing import List, Optional
import requests
from requests.exceptions import HTTPError
from src.main.python.preponderous.viron.models.grid import Grid

logger = logging.getLogger(__name__)

class GridService:
    def __init__(self, viron_host: str, viron_port: int):
        self.viron_host = viron_host
        self.viron_port = viron_port

    def get_base_url(self) -> str:
        return f"{self.viron_host}:{self.viron_port}/api/v1/grids"

    def get_all_grids(self) -> List[Grid]:
        response = requests.get(self.get_base_url())
        response.raise_for_status()
        return [Grid(**grid) for grid in response.json()]

    def get_grid_by_id(self, grid_id: int) -> Optional[Grid]:
        response = requests.get(f"{self.get_base_url()}/{grid_id}")
        response.raise_for_status()
        return Grid(**response.json())

    def get_grids_in_environment(self, environment_id: int) -> List[Grid]:
        response = requests.get(f"{self.get_base_url()}/environment/{environment_id}")
        response.raise_for_status()
        return [Grid(**grid) for grid in response.json()]

    def get_grid_of_entity(self, entity_id: int) -> Optional[Grid]:
        response = requests.get(f"{self.get_base_url()}/entity/{entity_id}")
        response.raise_for_status()
        return Grid(**response.json())